pkg/appfile/appfile.go

```diff
@@ -269,16 +273,27 @@ func generateComponentFromCUEModule(wl *Workload, appName, revision, ns string)
 }
 
 // evalWorkloadWithContext evaluate the workload's template to generate component and ACComponent
-func evalWorkloadWithContext(pCtx process.Context, wl *Workload, appName, compName string) (*v1alpha2.Component, *v1alpha2.ApplicationConfigurationComponent, error) {
+func evalWorkloadWithContext(pCtx process.Context, wl *Workload, ns, appName, compName string) (*v1alpha2.Component, *v1alpha2.ApplicationConfigurationComponent, error) {
+	var (
+		commonLabels      map[string]string
+		componentWorkload *unstructured.Unstructured
+		err               error
+	)
 	base, assists := pCtx.Output()
-	componentWorkload, err := base.Unstructured()
-	if err != nil {
-		return nil, nil, errors.Wrapf(err, "evaluate base template component=%s app=%s", compName, appName)
+	// if componentWorkload is nil, generated it based on Cue, or directly set it in Terraform Schematic case
+	if wl.CapabilityCategory == types.TerraformCategory {
+		componentWorkload, err = generateTerraformConfigurationWorkload(wl, ns)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to generate Terraform Configuration workload for workload %s", wl.Name)
+		}
```

Comment:
Maybe consider moving this logic into the switch block in line 190 and put it into an independent switch case (`case types.TerraformCategory`) out of default. That can make TerraforCategory more like (indeed it is) a parallel concept as CUE\Kube\Helm category.
Furthermore, just re-use `generateComponentFromCUEModule` in  `generateTerraformConfigurationWorkload` directly in the same way as  `generateComponentFromKubeModule` . The benefit is that you can avoid involving any logic about `category` into pkg `dsl/definition` and keep `AbstractEngine` interface unchanged. 
An additional benefit is that you can make Terraform-based to use CUE-based traits seamlessly, even though it may be not your original intention...

Reply:
agree with @captainroy-hy

Reply:
@captainroy-hy Thanks for your advice.
Yes, It looks better to put the code in the switch logic of L190 at the first glance, for which I thought it carefully when implementing it. And I looked through your implementation on `kube`. 
I wanted the code to be as simple as possible, so I chose not to do anything with CUE. This why I have to add one more parameter `category` in `AbstractEngine` to escape generating `outputs` and `output` by Cue lib.

If I have to implement it by CUE, I should follow your solution on `kube` module.